core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java

```diff
@@ -1250,9 +1249,9 @@ protected Scope visitTable(Table table, Optional<Scope> scope)
             ImmutableMap.Builder<ColumnHandle, Field> fields = ImmutableMap.builder();
 
             // TODO: discover columns lazily based on where they are needed (to support connectors that can't enumerate all tables)
-            TableMetadata tableMetadata = metadata.getTableMetadata(session, tableHandle);
             Map<String, ColumnHandle> columnHandles = metadata.getColumnHandles(session, tableHandle);
-            for (ColumnMetadata column : tableMetadata.getColumns()) {
+            for (ColumnHandle columnHandle : columnHandles.values()) {
+                ColumnMetadata column = metadata.getColumnMetadata(session, tableHandle, columnHandle);
```

Comment:
Does it usually call any metadata service to get column metadata? Please check `HiveMetadata` and `JdbcMetadata`

Reply:
Do we need to maintain the order of the columns ?

Reply:
> Do we need to maintain the order of the columns ?

Yes. And this requires that `getColumnMetadata` returns ImmutableMap or LinkedHashMap. Some connectors even rely on this. 
`io.trino.plugin.mongodb.MongoMetadata#getTableMetadata(io.trino.spi.connector.ConnectorSession, io.trino.spi.connector.SchemaTableName) `
I've one idea to modify returned type of io.trino.spi.connector.ConnectorMetadata#getColumnHandles from Map to ImmutableMap or at least update javadoc.
I'm investigating it further.

Reply:
there are even test failures for some other connectors:
```
Error:  blackHoleConnectorUsage(io.trino.plugin.blackhole.TestBlackHoleSmoke)  Time elapsed: 6.063 s  <<< FAILURE!
java.lang.RuntimeException: Insert query has mismatched column types: Table: [bigint, bigint, varchar(25), varchar(152)], Query: [bigint, varchar(25), bigint, varchar(152)]

Caused by: io.trino.spi.TrinoException: Insert query has mismatched column types: Table: [bigint, bigint, varchar(25), varchar(152)], Query: [bigint, varchar(25), bigint, varchar(152)]
	at io.trino.sql.analyzer.SemanticExceptions.semanticException(SemanticExceptions.java:48)
	at io.trino.sql.analyzer.SemanticExceptions.semanticException(SemanticExceptions.java:43)
	at io.trino.sql.analyzer.StatementAnalyzer$Visitor.visitInsert(StatementAnalyzer.java:461)
```

Reply:
> I've one idea to modify returned type of io.trino.spi.connector.ConnectorMetadata#getColumnHandles from Map to ImmutableMap or at least update javadoc.

The contract of this method is to return mapping, not to maintain the order, so modifying return type won't help unless you make sure that the map is always built accordingly to the expected order.
It seems that currently there is no other way to get ordered columns than via table metadata.

Reply:
> The contract of this method is to return mapping, not to maintain the order

The former does not forbid the latter.

> unless you make sure that the map is always built accordingly to the expected order.

Exactly. 

> modify returned type of io.trino.spi.connector.ConnectorMetadata#getColumnHandles from Map to ImmutableMap

It will require adding 
```
        <dependency>
            <groupId>com.google.guava</groupId>
            <artifactId>guava</artifactId>
        </dependency>
```
to
`trino/core/trino-spi/pom.xml`

I'll copy javadoc for `com.google.common.collect.ImmutableCollection` here:

> "Interfaces", not implementations
These are classes instead of interfaces to prevent external subtyping, but should be thought of as interfaces in every important sense. Each public class such as ImmutableSet is a type offering meaningful behavioral guarantees. This is substantially different from the case of (say) HashSet, which is an implementation, with semantics that were largely defined by its supertype.
For field types and method return types, you should generally use the immutable type (such as ImmutableList) instead of the general collection interface type (such as List). This communicates to your callers all of the semantic guarantees listed above, which is almost always very useful information.

Reply:
We can't add Guava as a dependency to the SPI. Different plugins may depend on different versions of Guava that are incompatible with each other.

Reply:
What if we could have an API that only fetches the `List<ColumnMetadata>` -> default implementation would be based on `ConnectorMetadata#getColumns` and each Connector could have it own implementation if needs to lazily load the properties

Reply:
> Does it usually call any metadata service to get column metadata? Please check HiveMetadata and JdbcMetadata

No. More detailed answer is here: https://github.com/trinodb/trino/pull/6745#discussion_r566207749